domain/schema/model/sql/0020-network.sql

```diff
@@ -23,3 +23,189 @@ CREATE TABLE cloud_container (
+-- it is considered public. A private network address is either
+-- specific to the cloud or cloud subnet a machine belongs to,
+-- or to the machine itself for containers.
+CREATE TABLE ip_address_scope (
+    id INT PRIMARY KEY,
+    name TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_ip_address_scope_name
+ON ip_address_scope (name);
+
+INSERT INTO ip_address_scope VALUES
+(0, 'unknown'),
+(1, 'public'),
+(2, 'local-cloud'),
+(3, 'local-machine'),
+(4, 'link-local');
+
+-- ip_address_config_type defines valid network
+-- link configuration types.
+CREATE TABLE ip_address_config_type (
+    id INT PRIMARY KEY,
+    name TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_ip_address_config_type_name
+ON ip_address_config_type (name);
+
+INSERT INTO ip_address_config_type VALUES
+(0, 'unknown'),
+(1, 'dhcp'),
+(2, 'static'),
+(3, 'manual'),
+(4, 'loopback');
+
+CREATE TABLE network_address (
+    uuid TEXT NOT NULL PRIMARY KEY,
+    origin_id INT NOT NULL,
+    scope_id INT NOT NULL,
+
+    CONSTRAINT fk_ip_address_origin
+    FOREIGN KEY (origin_id)
+    REFERENCES ip_address_origin (id),
+    CONSTRAINT fk_ip_address_scope
+    FOREIGN KEY (scope_id)
+    REFERENCES ip_address_scope (id)
+);
+
+CREATE TABLE ip_address (
+    address_uuid TEXT NOT NULL PRIMARY KEY,
+    -- The value of the configured IP address.
+    -- e.g. 192.168.1.2 or 2001:db8::1.
+    address_value TEXT NOT NULL,
+    type_id INT NOT NULL,
+    config_type_id INT NOT NULL,
+
+    CONSTRAINT chk_address_type
+    CHECK (type_id != 0),
+    CONSTRAINT fk_ip_address_address_uuid
+    FOREIGN KEY (address_uuid)
+    REFERENCES network_address (uuid),
+    CONSTRAINT fk_ip_address_type
+    FOREIGN KEY (type_id)
+    REFERENCES ip_address_type (id),
+    CONSTRAINT fk_ip_address_config_type
+    FOREIGN KEY (config_type_id)
+    REFERENCES ip_address_config_type (id)
+);
+
+CREATE TABLE fqdn_address (
+    address_uuid TEXT NOT NULL PRIMARY KEY,
+    hostname TEXT NOT NULL,
+
+    CONSTRAINT fk_fqdn_address_address_uuid
+    FOREIGN KEY (address_uuid)
+    REFERENCES network_address (uuid)
+);
+
+CREATE VIEW v_address AS
+SELECT
+    ipa.address_value,
+    na.type_id,
+    na.config_type_id,
+    na.origin_id,
+    na.scope_id
+FROM ip_address AS ipa
+INNER JOIN network_address AS na ON ipa.address_uuid = na.uuid
+UNION
+SELECT
+    fa.hostname AS address_value,
+    -- FQDN address type is always "hostname".
+    0 AS type_id,
+    -- FQDN address config type is always "manual".
+    3 AS config_type_id,
+    na.origin_id,
+    na.scope_id
+FROM fqdn_address AS fa
+INNER JOIN network_address AS na ON fa.address_uuid = na.uuid;
+
+CREATE TABLE ip_address_provider (
+    -- a provider-specific ID of the IP address.
+    provider_id TEXT NOT NULL PRIMARY KEY,
+    ip_address_uuid TEXT NOT NULL,
+    CONSTRAINT fk_provider_ip_address_ip_address_uuid
+    FOREIGN KEY (ip_address_uuid)
+    REFERENCES ip_address (uuid)
+);
+
+CREATE TABLE ip_address_space (
```

Comment:
I think we want to keep associating addresses with spaces via subnet. There are two reasons for this:
- The schema as posited here allows for an address to be in a different space to its subnet.
- Ultimately we really want to indicate that a _device_ is in a space rather than an (L3) address.